libraries/AP_Notify/Buzzer.cpp

```diff
@@ -178,7 +178,7 @@ void Buzzer::update()
     // check ekf bad
     if (_flags.ekf_bad != AP_Notify::flags.ekf_bad) {
         _flags.ekf_bad = AP_Notify::flags.ekf_bad;
-        if (_flags.ekf_bad) {
+        if (_flags.ekf_bad && AP_Notify::flags.armed && !hal.gpio->usb_connected()) {
```

Comment:
Given what was talked in the partners call and what Plane has done recently, I think the usb_connected part shouldn't be here.

Reply:
I think that's a behaviour change which can be part of a different PR?

I think we should systematically go through and remove special cases based on being connected via USB - apparently we're getting happier about people doing it.

Reply:
The behavior change is introduced by this PR... I'm not going to fuss on this because Plane doesn't set ekf_bad for now, but I still feel we shouldn't put the USB check here. If people really have a problem with buzzer going on when EKF goes bad and you are connected to USB and armed, then disabling or disconnecting the buzzer is always an option. Another option is to put this behavior in Copter and when USB is connected we don't do the EKF failsafe checks (basically what is there now, but instead just for USB).

Reply:
Honestly, as far as I know most of our users do not use a buzzer, so this is not a big deal. However, I think the armed and usb connected checks belong in Copter where the ekf_bad flag is set.

Reply:
As far as I know the ekf-bad flag should *really* still be set, even if you have USB connected.

The intent of the usb check is to silence the buzzer if you are just debugging the device while it is sitting on your desk - but people are increasingly flying with USB connected.

Thus, the "correct" place for the USB check is in the notify library, in the buzzer code.

I think everyone's pretty much agreed we need to stop special-casing when we're on USB.  But that's a topic for another PR (and probably a dev call!)  We now have a parameter which lets you disable the buzzer, which we didn't have when the on-USB hacks here went in.

Reply:
@peterbarker 

> As far as I know the ekf-bad flag should really still be set, even if you have USB connected.

And Sub sets it. The issue is that, with this change, Sub users won't get buzzer playing when EKF is bad, contrary to current behaviour.

> The intent of the usb check is to silence the buzzer if you are just debugging the device while it is sitting on your desk - but people are increasingly flying with USB connected.

Yes, but Sub doesn't have that behaviour now, which means this introduces a new behaviour for Sub that I think we are trying to remove.

> We now have a parameter which lets you disable the buzzer,

Yes, we do have that parameter which is why I think the USB check shouldn't be here.

>  which we didn't have when the on-USB hacks here went in.

But Sub doesn't have those USB hacks, this PR is what is introducing them to Sub in a moment where it isn't needed because we have the buzzer disable parameter.

I feel like I was a bit redundant above, but I feel like you aren't getting that this PR introduces new USB behaviour to Sub - and it's actually a behaviour we are seeking to remove from our code.

Reply:
> Thus, the "correct" place for the USB check is in the notify library, in the buzzer code.

Ok, I see this is the right place now. I think the USB check should be removed. Armed is fine, but I don't know if that's really desirable either.